File: built-in plugins dsp/mb-time-beatrepeat.cpp
/**
 * MB Beat Repeat
 * Category : effect
 * Type     : delay
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Capture and repeat beats with pitch and decay control
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_TIME_BEATREPEAT_H
#define MB_TIME_BEATREPEAT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbTimeBeatrepeat : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-time-beatrepeat";
    static constexpr const char* PLUGIN_NAME    = "MB Beat Repeat";
    static constexpr const char* PLUGIN_TYPE    = "delay";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float interval = 4f;  // range [1, 16]
    float repeats = 4f;  // range [1, 32]
    float decay = 0.5f;  // range [0, 1]
    float pitchShift = 0f;  // range [-12, 12]
    float mix = 1f;  // range [0, 1]
    };

    MbTimeBeatrepeat() = default;
    ~MbTimeBeatrepeat() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.interval = std::clamp(params.interval, 1f, 16f);
        params.repeats = std::clamp(params.repeats, 1f, 32f);
        params.decay = std::clamp(params.decay, 0f, 1f);
        params.pitchShift = std::clamp(params.pitchShift, -12f, 12f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Beat Repeat
        return input;
    }
};

#endif // MB_TIME_BEATREPEAT_H

File: built-in plugins dsp/mb-time-beatrepeat.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbTimeBeatrepeatPlugin: PluginDefinition = {
  id: "mb-time-beatrepeat",
  slug: "mb-time-beatrepeat",
  name: "MB Beat Repeat",
  category: "effect",
  type: "delay" as any,
  version: "1.0.0",
  description: "Capture and repeat beats with pitch and decay control",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "interval",
      name: "Interval",
      type: "float",
      defaultValue: 4,
      minValue: 1,
      maxValue: 16,
      automatable: true,
    },
    {
      id: "repeats",
      name: "Repeats",
      type: "float",
      defaultValue: 4,
      minValue: 1,
      maxValue: 32,
      automatable: true,
    },
    {
      id: "decay",
      name: "Decay",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "pitchShift",
      name: "Pitch Shift",
      type: "float",
      defaultValue: 0,
      minValue: -12,
      maxValue: 12,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 1,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { interval: 4, repeats: 4, decay: 0.5, pitchShift: 0, mix: 1 },
};

export default MbTimeBeatrepeatPlugin;
